snuba/datasets/tags_dataset.py

```diff
@@ -0,0 +1,140 @@
+import re
+from typing import Any, Mapping, MutableMapping, Set, Union
+
+from snuba import state
+from snuba.clickhouse.columns import ColumnSet
+from snuba.util import (
+    alias_expr,
+    all_referenced_columns,
+    escape_literal,
+    escape_col,
+)
+
+# A column name like "tags[url]"
+NESTED_COL_EXPR_RE = re.compile(r'^(tags|contexts)\[([a-zA-Z0-9_\.:-]+)\]$')
+
+
+class TagColumnProcessor:
+    """
+    Provides the query processing features for tables that have
+    tags and contexts nested columns. This simply extracts these
+    features originally developed in the Events dataset.
+    """
+
+    def __init__(self,
+        columns: ColumnSet,
+        promoted_columns: Mapping[str, Set[str]],
+        column_tag_map: Mapping[str, Mapping[str, str]]
+    ) -> None:
+        # The ColumnSet of the dataset. Used to format promoted
+        # columns with the right type.
+        self.__columns = columns
+        # Keeps a dictionary of promoted columns. the key of the mapping
+        # can be 'tags' or 'contexts'. The values is a set of flattened
+        # columns.
+        self.__promoted_columns = promoted_columns
+        # A mapping between column representing promoted tags and the
+        # correspinding tag.
+        self.__column_tag_map = column_tag_map
+
+    def process_tags_expression(self,
+        column_name: str,
+        body: MutableMapping[str, Any],
+    ) -> Union[None, Any]:
+        """
+        This method resolves the tag or context processes it and formats
+        the column expression for the query. It is supposed to be called
+        by column_expr methods in the datasets.
+
+        It returns None if the column is not a tag or context.
+        """
+        if NESTED_COL_EXPR_RE.match(column_name):
+            return self.__tag_expr(column_name)
+        elif column_name in ['tags_key', 'tags_value']:
```

Comment:
How come we do not need to have a similar check for contexts_key and contexts_value?

Reply:
That's an interesting question. This is how it works on events. I don't know if there is a reason for not doing that on contexts or if it was just never asked for.